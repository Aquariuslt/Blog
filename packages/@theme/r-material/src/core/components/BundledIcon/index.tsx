import * as React from 'react';

import { ChevronLeft, ExpandLess, ExpandMore, Menu, StarBorder } from '@material-ui/icons';

interface BundleIconProps {
  type: string;
  fallback?: string;
}

const FALLBACK_TYPE = 'menu';

/**
 * Provides a component with bundled icon types and fallback options
 */
const BundledIcon: React.ComponentType<BundleIconProps> = (props: BundleIconProps): JSX.Element => {
  const bundledMap = {
    menu: <Menu />,
    chevronLeft: <ChevronLeft />,
    starBorder: <StarBorder />,
    expandLess: <ExpandLess />,
    expandMore: <ExpandMore />
  };

  const detectIconType = (): string => {
    if (Object.keys(bundledMap).includes(props.type)) {
      return props.type;
    }
    return props.fallback || FALLBACK_TYPE;
  };

  return bundledMap[detectIconType()];
};

export default BundledIcon;
