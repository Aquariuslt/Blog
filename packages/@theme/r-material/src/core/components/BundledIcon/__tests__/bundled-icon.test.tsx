import * as React from 'react';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

import Menu from '@material-ui/icons/Menu';
import BundledIcon from '../';

Enzyme.configure({ adapter: new Adapter() });

describe('@theme/r-material: core/components: BundledIcon', (): void => {
  it('# should mount correct bundled types correctly', (): void => {
    const mockType = 'expandLess';
    const wrapper = Enzyme.shallow(<BundledIcon type={mockType} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('# should mount with same dom with material-ui/icon', (): void => {
    const mockType = 'menu';

    const muiMenuWrapper = Enzyme.shallow(<Menu />);
    const bundleIconWrapper = Enzyme.shallow(<BundledIcon type={mockType} />);

    expect(bundleIconWrapper.html()).toEqual(muiMenuWrapper.html());
  });

  it('# should fallback to menu when type is not in bundled map', (): void => {
    const mockType = 'somehow';
    const muiMenuWrapper = Enzyme.shallow(<Menu />);
    const fallbackBundleIconWrapper = Enzyme.shallow(<BundledIcon type={mockType} />);

    expect(fallbackBundleIconWrapper.html()).toEqual(muiMenuWrapper.html());
  });
});
