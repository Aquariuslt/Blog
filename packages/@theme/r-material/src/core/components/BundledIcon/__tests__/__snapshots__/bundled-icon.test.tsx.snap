// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@theme/r-material: core/components: BundledIcon # should mount correct bundled types correctly 1`] = `<ExpandLessIcon />`;
